### ISSUE TITLE
Show new page with Information how to contribute when no readme or test files found

### ISSUE THREAD

**Issue by GustavoCaso** [exercism/todo#174]
With the api v3 we get the readme and test files from an exercises, there are cases where they are empty, we should display a new page giving proper information how to contribute

**Comment by kytrinyx** [comment]
More context:

This is in the https://github.com/exercism/exercism.io repository. The routes are in `app/routes/metadata.rb`.

The route should set a status 404 in the case where it's not found.

"how to contribute" would be to suggest that they implement the missing problem in the track.

The contributing guide has instructions for how to do that. https://github.com/exercism/x-common/blob/master/CONTRIBUTING.md#porting-an-exercise-to-another-language-track

We can also tell them what the repository url is.

**Comment by GustavoCaso** [comment]
Thanks

**Comment by heatherlim** [comment]
@GustavoCaso @kytrinyx Hey guys has the path to metadata been moved?

**Comment by kytrinyx** [comment]
@heatherlim Yes, good catch, these are now at the bottom of this file: app/routes/exercises.rb

**Comment by scottwhudson** [comment]
Is this back up for grabs @kytrinyx ?

**Comment by kytrinyx** [comment]
Yeah, it looks like it. I kind of want to make a policy that if we don't hear back within a certain amount of time it automatically goes back up for grabs.

**Comment by scottwhudson** [comment]
This is in an unintended issue reference due to the conversation in the other repo, the referenced branch isn't intended to fix this issue

**Comment by scottwhudson** [comment]
@kytrinyx by any chance, do you know if there Is there a track that's missing a bunch of readme's or test files? I'm striking out trying to find one

**Comment by Insti** [comment]
@scottwhudson There are a lot of exercises that are missing common test data: https://github.com/exercism/todo/issues?q=is%3Aissue+is%3Aopen+label%3Ashared-test-data

**Comment by scottwhudson** [comment]
Ah ok, so I may have made an incorrect assumption, which is that some of the existing readme/test file links are broken because they don't yet exist.  When I click through the available exercises in each track though, I don't find that to be the case.  

Will an exercise only display in the language track if a README and test file exist? For example, all of the current [ruby exercises](http://exercism.io/languages/ruby) have READMEs and test files, but if I try an exercise that doesn't exist in the track (i.e. [zipper](http://exercism.io/exercises/ruby/zipper/readme) doesn't exist in ruby), I get the behavior I'm looking for in the routes (redirect to root).  But I have to manually enter that path, there's no button for zipper in the ruby track

If the goal here is to make all exercises available in all languages, shouldn't a button exist for every exercise in every track that redirects to a contribution page if necessary? I'm just not understanding this from a UX perspective.

**Comment by Insti** [comment]
I'm not entirely sure what this issue is asking either.

> If the goal here is to make all exercises available in all languages, shouldn't a button exist for every exercise in every track that redirects to a contribution page if necessary?

This page: http://exercism.io/languages/ruby/contribute (hidden 3 or 4 levels deep on the site) lists the problems still to be implemented.

I'm not sure people hand-editing urls looking for unimplemented exercises is a useful use-case to bother catering to.

Or maybe it is the intention that the CLI would return a useful contributing link rather than the current 404 message?
> $ exercism fetch ruby zipper
> 2016/07/26 22:03:13 unable to fetch problems (HTTP: 404) - No implementation of 'zipper' in track 'ruby'

**Comment by Insti** [comment]
@scottwhudson  I noticed after writing that last post that the porting instructions are out of date so I've added issue #182 maybe you'd like to work on that one while you're waiting for some answers here.

**Comment by scottwhudson** [comment]
Ah, that CLI use-case makes a lot of sense, I didn't get the impression that was a goal since the CLI code lives outside of the sinatra app if I'm not mistaken. But I totally agree.

And sure @Insti, I gotcha. 👍

The back end concerns make sense, but if someone can give a basic set of user requirements on how this feature should work, that would be soopah.

**Comment by scottwhudson** [comment]
I started messing around with this idea while discussing the issue above.  If the ultimate goal here is to create a platform that encourages uniformity across language tracks, why not incorporate visual cues in the track view to denote exercises that still need work in order to get ported? I tweaked some basic CSS in the browser to get this initial concept:

<img width="994" alt="screen shot 2016-07-26 at 4 35 33 pm" src="https://cloud.githubusercontent.com/assets/6406203/17156358/0943525e-534f-11e6-96b6-015985a0ec75.png">


'In-progress' exercises would be pushed to the bottom of the list, have a distinct background color, and contain contribution info when opened (perhaps even specifically "this needs a README!" etc.) So all exercises would display for all tracks, and the incomplete exercises would appear this way. 

Fair warning, I'm not quite familiar enough with the codebase yet to determine how much work this would be.  This is just my WAG as to what a UI change would look like for this feature.

**Comment by Insti** [comment]
I like the idea of making exercises that need contribution more visible.
(Although Ruby may not be the best example to use for this since it already has many.)

But I'm not sure they should be included with the list of available exercises.
Perhaps adding some text to the bottom of that page that says something like:
"Solved all the exercises? Help add new ones <link to contribute page>"

Thanks for the mock-up, these are always helpful.

**Comment by kytrinyx** [comment]
Thanks so much for continuing this discussion, @scottwhudson!

You're right, this issue is a bit fuzzy. Let's try to clarify.

1. The only reason why you would get to a URL for a README or test file that doesn't exist, is if someone hand-crafts that URL. We should simply return a 404 on the website.
2. If someone tries fetching an exercise that exists but is unimplemented, then we should probably go with @Insti's suggestion of having a useful error message for that case.
3. It would be lovely to figure out how to encourage people to help port exercises, like with your mock-up above. Let's make that a separate issue over in github.com/exercism/discussions/issues so that we can hash out some ideas, and then we can create focused issues once we've figured it out.

**Comment by kytrinyx** [comment]
Oh, and 4. We have started working on extracting a canonical dataset for exercises, and there are many exercises that don't have this data yet. We could encourage people to do that, as well.

I think this issue should be about returning the 404 correctly in the case of a README or tests that don't exist when someone goes to the URL in question.

**Comment by scottwhudson** [comment]
Sorry, another noob question: should this fix be implemented in the CLI repo or exercism.io repo? It seems straightforward to just output a message with a link to the generic contribution page in the CLI without touching the exercism API, but how would you guys want to do it?

**Comment by kytrinyx** [comment]
Don't apologize—it's always complicated when things are in a bajillion different places. Also we've been talking about several changes.

1. If you try to go to a README or test suite on the website and that doesn't exist, return a 404.
   _Currently, if you do this, you get redirected back to the home page with an error message. This seems fine._
2. If someone tries fetching an exercise that exists but is unimplemented, then we should have a useful error message for that case.
    _We could do either. If we change the API, then this should JustWork(tm) in the CLI, because it will display whatever comes back in the "error" field in the case of an error. If we change the CLI, then we would have to make sure that the API returns the correct error code anyway (which it may or may not)._

For the two other issues, would you mind opening separate discussions in the discussions repository? I think it would be worth hashing out some ideas about how to tackle them.

**Comment by scottwhudson** [comment]
I just referenced a PR for this issue. I kept running into an issue where my local API was running v1 routes instead of v3, which appears to be incorrect.  Is there a config step I'm missing somewhere or something?

**Comment by scottwhudson** [comment]
@kytrinyx I just opened an issue in discussions about the UI idea I had, what is the other issue I should open?

**Comment by kytrinyx** [comment]
> what is the other issue I should open?

I think I was just confused. It's just this one. Thanks!
